<div style="background-color: #f8f9fa; padding: 20px; border-radius: 10px; box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);">
  <div style="display: flex; justify-content: space-between; align-items: center;">
    <img src="https://raw.githubusercontent.com/carlosfab/blog-sigmoidal/main/_assets/logo_sigmoidal.png" alt="Sigmoidal" style="height: 60px;">
    <div style="text-align: right;">
      <p style="margin: 0; font-size: 14px; color: #6c757d;"><strong>Carlos Melo</strong> · Sigmoidal</p>
      <p style="margin: 0; font-size: 12px; color: #6c757d;">Cientista de Dados · Visão Computacional</p>
    </div>
  </div>
</div>


# Introdução ao Pandas para Análise Exploratória

Material de apoio do artigo [Introdução ao Pandas para Análise Exploratória](https://sigmoidal.ai) no blog da Sigmoidal.

A ideia aqui é **prática**: você vai carregar um arquivo `csv` real, entender o que tem dentro dele, calcular estatísticas, criar um `DataFrame` do zero e tratar valores ausentes. Tudo o que um cientista de dados faz nos primeiros minutos de qualquer projeto.

> O notebook foi pensado para rodar de ponta a ponta no Google Colab gratuito, sem dependências além de `pandas` e `numpy`.


## 1. Importando as bibliotecas

Pandas costuma vir junto com qualquer ambiente Python científico (Anaconda, Colab, Kaggle). Se for instalar do zero na sua máquina, basta `pip install pandas`.

Por convenção, importa-se com o apelido `pd`. Vamos importar também o `numpy`, que será usado mais à frente.


In [1]:
import pandas as pd
import numpy as np

print(f"pandas: {pd.__version__}")
print(f"numpy:  {np.__version__}")


pandas: 2.3.3
numpy:  1.26.4


## 2. Carregando dados de um arquivo CSV

A função `pd.read_csv()` aceita tanto um caminho local quanto uma URL. Aqui vamos carregar o histórico de cotações da ação ordinária do **Banco do Brasil (BBAS3)**, hospedado num repositório público no GitHub.


In [2]:
url_bbas3 = "https://raw.githubusercontent.com/carlosfab/dsnp2/master/datasets/BBAS3.SA.csv"
df = pd.read_csv(url_bbas3)
df.head()


,Date,Open,High,Low,Close,Adj Close,Volume
0,2019-05-06,49.470001,49.669998,49.180000,49.369999,46.886795,6314800.0
1,2019-05-07,49.080002,49.680000,48.220001,49.490002,47.000763,10846100.0
2,2019-05-08,49.500000,50.840000,49.500000,50.590000,48.045422,14459600.0
3,2019-05-09,50.910000,51.259998,50.119999,51.029999,48.463291,15377500.0
4,2019-05-10,51.099998,51.189999,49.860001,50.160000,47.637051,8975300.0


Pandas reconhece muitos outros formatos: JSON, Parquet, HDF, Excel, SQL e por aí vai. A lógica é sempre a mesma, uma função `pd.read_*()` devolvendo um `DataFrame`.


## 3. Inspecionando o DataFrame

Antes de qualquer análise, é preciso ganhar **consciência situacional** sobre o dado: quantas linhas, quantas colunas, que tipos, se tem valor faltando.


In [3]:
# tipo do objeto
type(df)


pandas.core.frame.DataFrame

In [4]:
# tamanho (linhas, colunas)
df.shape


(248, 7)

In [5]:
# nomes das colunas
df.columns


Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

In [6]:
# 5 primeiras linhas
df.head()


,Date,Open,High,Low,Close,Adj Close,Volume
0,2019-05-06,49.470001,49.669998,49.180000,49.369999,46.886795,6314800.0
1,2019-05-07,49.080002,49.680000,48.220001,49.490002,47.000763,10846100.0
2,2019-05-08,49.500000,50.840000,49.500000,50.590000,48.045422,14459600.0
3,2019-05-09,50.910000,51.259998,50.119999,51.029999,48.463291,15377500.0
4,2019-05-10,51.099998,51.189999,49.860001,50.160000,47.637051,8975300.0


In [7]:
# 5 ultimas linhas
df.tail()


,Date,Open,High,Low,Close,Adj Close,Volume
243,2020-04-27,25.500000,26.200001,25.110001,25.690001,25.690001,32132800.0
244,2020-04-28,27.219999,29.250000,27.110001,29.139999,29.139999,43232700.0
245,2020-04-29,30.040001,30.209999,28.730000,29.559999,29.559999,24628700.0
246,2020-04-30,28.309999,28.780001,27.850000,28.500000,28.500000,23935300.0
247,2020-05-04,27.370001,27.760000,26.990000,27.760000,27.760000,16586000.0


In [8]:
# tipos de cada coluna
df.dtypes


Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume       float64
dtype: object

O método `.info()` traz tudo de uma vez: nome da coluna, contagem de valores não nulos e tipo de cada coluna. Repare que a coluna `Open` tem 247 valores não nulos contra 248 do `Date`, ou seja, há uma linha com numéricos faltando. Esse tipo de pista aparece já na primeira inspeção.


In [9]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       248 non-null    object 
 1   Open       247 non-null    float64
 2   High       247 non-null    float64
 3   Low        247 non-null    float64
 4   Close      247 non-null    float64
 5   Adj Close  247 non-null    float64
 6   Volume     247 non-null    float64
dtypes: float64(6), object(1)
memory usage: 13.7+ KB


## 4. DataFrame vs Series

Um `DataFrame` é a tabela inteira. Uma `Series` é uma coluna isolada. Dá para selecionar uma coluna por colchetes ou por ponto.


In [10]:
# selecionar uma coluna por colchetes
df["High"].head()


0    49.669998
1    49.680000
2    50.840000
3    51.259998
4    51.189999
Name: High, dtype: float64

In [11]:
# mesma coluna, sintaxe alternativa
df.High.head()


0    49.669998
1    49.680000
2    50.840000
3    51.259998
4    51.189999
Name: High, dtype: float64

In [12]:
# tipo de uma coluna isolada
type(df["High"])


pandas.core.series.Series

## 5. Trabalhando com datas

Repare em `df.dtypes` que a coluna `Date` veio como `object`, ou seja, texto. Pandas não reconhece datas automaticamente, então é preciso converter explicitamente com `pd.to_datetime()`.


In [13]:
df.Date.head()


0    2019-05-06
1    2019-05-07
2    2019-05-08
3    2019-05-09
4    2019-05-10
Name: Date, dtype: object

In [14]:
df["Date"] = pd.to_datetime(df["Date"], format="%Y-%m-%d")
df.dtypes


Date         datetime64[ns]
Open                float64
High                float64
Low                 float64
Close               float64
Adj Close           float64
Volume              float64
dtype: object

Com a coluna no tipo `datetime64`, ganhamos o acessor `.dt`, que extrai componentes da data sem complicação.


In [15]:
df["Year"] = df.Date.dt.year
df["Month"] = df.Date.dt.month
df.head()


,Date,Open,High,Low,Close,Adj Close,Volume,Year,Month
0,2019-05-06,49.470001,49.669998,49.180000,49.369999,46.886795,6314800.0,2019,5
1,2019-05-07,49.080002,49.680000,48.220001,49.490002,47.000763,10846100.0,2019,5
2,2019-05-08,49.500000,50.840000,49.500000,50.590000,48.045422,14459600.0,2019,5
3,2019-05-09,50.910000,51.259998,50.119999,51.029999,48.463291,15377500.0,2019,5
4,2019-05-10,51.099998,51.189999,49.860001,50.160000,47.637051,8975300.0,2019,5


## 6. Estatísticas descritivas

O método `.describe()` traz contagem, média, desvio padrão, mínimo, máximo e os quartis (25%, 50%, 75%) de cada coluna numérica de uma só vez.


In [16]:
df.describe()


,Date,Open,High,Low,Close,Adj Close,Volume,Year,Month
count,248,247.000000,247.000000,247.000000,247.000000,247.000000,2.470000e+02,248.000000,248.000000
mean,2019-10-31 06:17:25.161290240,45.939960,46.576397,45.154615,45.838097,44.684901,1.405295e+07,2019.334677,6.487903
min,2019-05-06 00:00:00,24.200001,25.000000,21.910000,22.129999,22.129999,5.092200e+06,2019.000000,1.000000
25%,2019-08-01 18:00:00,45.425002,46.145001,44.774999,45.415001,44.072172,8.997050e+06,2019.000000,3.750000
50%,2019-10-28 12:00:00,48.139999,48.779999,47.560001,47.990002,46.766251,1.198600e+07,2019.000000,7.000000
75%,2020-01-30 06:00:00,50.535000,51.125000,49.785000,50.460001,49.133178,1.587985e+07,2020.000000,9.000000
max,2020-05-04 00:00:00,55.599998,55.700001,54.820000,55.180000,53.052986,7.397800e+07,2020.000000,12.000000
std,NaN,7.644318,7.475109,7.858680,7.704523,7.210382,8.021174e+06,0.472832,3.428839


Para uma coluna específica, dá para chamar cada estatística diretamente.


In [17]:
# media e mediana
print(f"media:   {df.High.mean():.2f}")
print(f"mediana: {df.High.median():.2f}")


media:   46.58
mediana: 48.78


In [18]:
# desvio padrao e variancia
print(f"std: {df.High.std():.2f}")
print(f"var: {df.High.var():.2f}")


std: 7.48
var: 55.88


In [19]:
# minimo e maximo
print(f"min: {df.High.min():.2f}")
print(f"max: {df.High.max():.2f}")


min: 25.00
max: 55.70


### Mediana é mais resistente a outliers

Considere um conjunto pequeno com um valor extremo: `[1, 1, 2, 4, 900]`. A média é puxada pelo `900`. A mediana ignora esse efeito.


In [20]:
amostra = pd.Series([1, 1, 2, 4, 900])
print(f"media:   {amostra.mean()}")
print(f"mediana: {amostra.median()}")


media:   181.6
mediana: 2.0


## 7. Criando DataFrames do zero

Nem sempre você parte de um `csv`. Em testes, simulações e exemplos didáticos, você vai querer construir um `DataFrame` na hora.

### A partir de um dicionário

A chave do dicionário vira o nome da coluna; o valor vira a coluna em si.


In [21]:
dados = {
    "nome": ["Carlos", "Theo", "Raquel", "Fernanda"],
    "idade": [35, 32, 15, 49],
    "cidade": ["Araraquara", "Belém", "Natal", "Curitiba"],
    "comprou": [True, False, False, True],
}

clientes = pd.DataFrame(dados)
clientes


,nome,idade,cidade,comprou
0,Carlos,35,Araraquara,True
1,Theo,32,Belém,False
2,Raquel,15,Natal,False
3,Fernanda,49,Curitiba,True


Posso definir o índice manualmente, por exemplo, usando o id de cada usuário.


In [22]:
clientes.index = [8712, 5831, 4421, 9873]
clientes


,nome,idade,cidade,comprou
8712,Carlos,35,Araraquara,True
5831,Theo,32,Belém,False
4421,Raquel,15,Natal,False
9873,Fernanda,49,Curitiba,True


### A partir de listas

Quando os dados estão em listas separadas, dá para passar tudo de uma vez como uma lista de listas.


In [23]:
linhas = [
    ["Carlos", 35, "Araraquara", True],
    ["Theo", 32, "Belém", False],
    ["Raquel", 15, "Natal", False],
    ["Fernanda", 49, "Curitiba", True],
]

clientes2 = pd.DataFrame(
    linhas,
    columns=["nome", "idade", "cidade", "comprou"],
)
clientes2


,nome,idade,cidade,comprou
0,Carlos,35,Araraquara,True
1,Theo,32,Belém,False
2,Raquel,15,Natal,False
3,Fernanda,49,Curitiba,True


### A partir de NumPy

Para simulações, é comum gerar dados aleatórios com `numpy` e empacotar num `DataFrame`.


In [24]:
rng = np.random.default_rng(seed=42)

amostras = pd.DataFrame({
    "altura_cm":  rng.normal(loc=170, scale=10, size=200),
    "peso_kg":    rng.normal(loc=70,  scale=12, size=200),
    "horas_sono": rng.normal(loc=7,   scale=1.2, size=200),
})

amostras.describe()


,altura_cm,peso_kg,horas_sono
count,200.000000,200.000000,200.000000
mean,169.695493,70.240931,6.918252
std,8.819329,12.234058,1.219808
min,148.679537,39.200099,3.442565
25%,163.434277,61.644245,6.062496
50%,169.481458,70.725396,6.905931
75%,175.350205,77.669621,7.675113
max,199.138625,104.860806,10.117208


## 8. Tratando valores ausentes

Dado real quase nunca vem completo. Para essa parte, vamos usar um dataset de **corridas de bicicleta compartilhada** (Brasília, 2018), com cerca de 287 mil registros e vários campos com nulos.


In [25]:
url_rides = "https://dl.dropboxusercontent.com/s/yyfeoxqw61o3iel/df_rides.csv"
rides = pd.read_csv(url_rides, parse_dates=["user_birthdate", "ride_date"])
rides.shape


(287322, 10)

In [26]:
rides.head()


,user_gender,user_birthdate,user_residence,ride_date,time_start,time_end,station_start,station_end,ride_duration,ride_late
0,M,1971-06-08,NaN,2018-01-01,06:05:18,06:21:33,11 - Rodoviária 2,41 - Instituto de Artes,16.250000,0.0
1,M,1989-02-11,DF,2018-01-01,06:27:01,06:32:17,26 - Ministério da Saude,28 - CNMP - Conselho Nacional do Ministério Pú...,5.266667,0.0
2,M,1968-07-19,NaN,2018-01-01,06:29:33,06:44:57,11 - Rodoviária 2,43 - Biblioteca Central,15.400000,0.0
3,M,1991-12-19,NaN,2018-01-01,06:53:53,06:59:45,10 - Ministério dos Transportes,6 - Rodoviária,5.866667,0.0
4,M,1969-03-03,DF,2018-01-01,06:58:56,17:40:04,15 - Brasil 21,11 - Rodoviária 2,641.133333,1.0


### Detectando valores ausentes

`isna()` devolve um `DataFrame` booleano (`True` onde o valor é nulo). Combinado com `.sum()`, conta nulos por coluna.


In [27]:
rides.isna().sum()


user_gender          396
user_birthdate         1
user_residence    179905
ride_date              0
time_start             0
time_end           43285
station_start          0
station_end            0
ride_duration      73174
ride_late          73174
dtype: int64

Mais de 60% dos registros não tem `user_residence` informada, e cerca de 43 mil corridas não tem `time_end` (provavelmente registros incompletos do app).


### Removendo linhas com nulos: `dropna`

`dropna()` apaga linhas que têm pelo menos um valor ausente. Use o argumento `subset` para olhar apenas determinadas colunas.


In [28]:
# antes
print(f"antes:  {len(rides):,} linhas")

rides_limpo = rides.dropna(subset=["user_residence"])

print(f"depois: {len(rides_limpo):,} linhas")


antes:  287,322 linhas
depois: 107,417 linhas


### Preenchendo nulos: `fillna`

Em muitos casos, jogar fora a linha inteira é caro demais. Uma alternativa pragmática é preencher: mediana para colunas numéricas, moda (valor mais frequente) ou uma categoria explícita para categóricas.


In [29]:
# preencher residencia ausente com a moda
moda_residencia = rides["user_residence"].mode()[0]
print(f"moda: {moda_residencia}")

rides["user_residence"] = rides["user_residence"].fillna(moda_residencia)
rides.isna().sum()


moda: DF


user_gender         396
user_birthdate        1
user_residence        0
ride_date             0
time_start            0
time_end          43285
station_start         0
station_end           0
ride_duration     73174
ride_late         73174
dtype: int64

> **Cuidado**: imputar com mediana ou moda é uma escolha pragmática quando você não tem informação a priori. Em pipelines de machine learning, isso vira parte de um `Pipeline` do `scikit-learn`, e é fundamental calcular estatísticas só no conjunto de treino para evitar **data leakage**.


## 9. Próximos passos

Você cobriu o essencial: **carregar, inspecionar, criar e tratar** `DataFrames`. Daqui pra frente, o caminho natural é:

- **Visualização** (histogramas, boxplots, scatter plots) com `pandas` e `matplotlib`.
- **Filtragem e agrupamento** (`.query()`, `.groupby()`, `.merge()`) para responder perguntas mais específicas sobre o dado.
- **Pipelines de machine learning** com `scikit-learn`, onde o tratamento de nulos vira um passo formal.

Esses tópicos virão em artigos seguintes do blog. Bons estudos!
